File: 103BinaryTreeZigzagLevelOrderTraversal.cpp
//1. Keep a bool zigzag value to see whether move l-r or r-l.
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> res;
        if(!root) return res;
        
        queue<TreeNode *> nodes;
        nodes.push(root);
        bool zigzag = true;
        while(!nodes.empty()){
            vector<int> level;
            int len = nodes.size();
            for(int i=0;i<len;i++){
                level.push_back(nodes.front()->val);
                if(nodes.front()->left) nodes.push(nodes.front()->left);
                if(nodes.front()->right) nodes.push(nodes.front()->right);
                nodes.pop();
            }
            if(!zigzag) reverse(level.begin(),level.end());
            res.push_back(level);
            zigzag = !zigzag;
        }
        return res;
    }
};

//2. Iteration
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> res;
        if(!root) return res;
        
        zigzagLevelOrder(root,1,res,true);
        return res;
    }
    
    void zigzagLevelOrder(TreeNode* root, int level, vector<vector<int>>& res, bool zigzag) {
        if(!root) return;
        
        if(level>res.size()) res.push_back(vector<int>());
        
        if(zigzag) res[level-1].push_back(root->val);
        else res[level-1].insert(res[level-1].begin(),root->val);
        
        zigzagLevelOrder(root->left,level+1,res,!zigzag);
        zigzagLevelOrder(root->right,level+1,res,!zigzag);
        
        
    }
};

File: Algo/Leetcode/103BinaryTreeZigzagLevelOrderTraversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        tmp = []
        res = []
        if root is None:
            return tmp

        toRight = True
        tmp.append(root)
        while len(tmp) > 0:
            count = len(tmp)
            level = []
            while count > 0:
                n = tmp[0]
                level.append(n.val)
                if n.left is not None:
                    tmp.append(n.left)
                if n.right is not None:
                    tmp.append(n.right)
                tmp.pop(0)
                count -= 1
            if not toRight:
                level = level[::-1]
            res.append(level)

            toRight = not toRight

        return res